File: src/test/java/com/github/thaynarasilvapinto/SimuladorBanco/ContaDonainTest.java
package com.github.thaynarasilvapinto.SimuladorBanco;

import com.github.thaynarasilvapinto.SimuladorBanco.domain.Cliente;
import com.github.thaynarasilvapinto.SimuladorBanco.domain.Conta;
import com.github.thaynarasilvapinto.SimuladorBanco.domain.Operacao;
import com.github.thaynarasilvapinto.SimuladorBanco.domain.TipoOperacao;
import org.junit.Before;
import org.junit.Test;

import static org.junit.Assert.assertEquals;


public class ContaDonainTest {

    private Cliente joao;
    private Cliente maria;
    private Conta joaoConta;
    private Conta mariaConta;

    private Operacao operacaoDepositoJoao;
    private Operacao operacaoSaqueJoao;
    private Operacao operacaoTransferencia;

    @Before
    public void setup() {
        mariaConta = new Conta(0.00);
        joaoConta = new Conta(0.00);
        this.joao = new Cliente("João Pedro da Silva", "15142542675", joaoConta);
        this.maria = new Cliente("Maria Abadia de Oliveira", "17708289667", mariaConta);


        this.operacaoDepositoJoao = new Operacao(joaoConta, joaoConta, 200.00, TipoOperacao.DEPOSITO);
        this.operacaoSaqueJoao = new Operacao(joaoConta, joaoConta, 100.00, TipoOperacao.SAQUE);
        this.operacaoTransferencia = new Operacao(joaoConta, mariaConta, 100, TipoOperacao.TRANSFERENCIA);
    }

    @Test
    public void deveRealizarDeposito() {
        joao.getConta().deposito(operacaoDepositoJoao);

        double saldoEsperado = 200.00;
        double saldoAtual = joao.getConta().getSaldo();

        assertEquals(saldoEsperado, saldoAtual, 0.00001);
    }

    @Test
    public void deveRealizarSaque() {
        double saldoEsperado = 100.00;
        double saldoAtual;

        joao.getConta().deposito(operacaoDepositoJoao);
        joao.getConta().saque(operacaoSaqueJoao);

        saldoAtual = joao.getConta().getSaldo();

        assertEquals(saldoEsperado, saldoAtual, 0.00001);
    }

    @Test
    public void naoDeveRealizarSaqueComSaldo0EValorMaiorQueOSaldo() {
        double saldoEsperado = 0.0;
        double saldoAtual;
        Operacao statusDaOperacaoSaque = joao.getConta().saque(operacaoSaqueJoao);
        Operacao statusEsperadoDaOpercaoSaque = null;

        saldoAtual = joao.getConta().getSaldo();
        assertEquals(statusDaOperacaoSaque, statusEsperadoDaOpercaoSaque);
        assertEquals(saldoEsperado, saldoAtual, 0.00001);
    }

    @Test
    public void deveRealizarTransferencia() {
        double saldoEsperadoJoao = 100;
        double saldoEsperadoMaria = 100;
        double saldoAtualJoao;
        double saldoAtualMaria;
        joao.getConta().deposito(operacaoDepositoJoao);
        joao.getConta().Transferencia(maria.getConta(), operacaoTransferencia);
        saldoAtualJoao = joao.getConta().getSaldo();
        saldoAtualMaria = maria.getConta().getSaldo();

        assertEquals(saldoEsperadoJoao, saldoAtualJoao, 0.00001);
        assertEquals(saldoEsperadoMaria, saldoAtualMaria, 0.00001);
    }

    @Test
    public void naoDeveRealizarTransacaoDeValorMaisAltoDoQueOSaldo() {
        Operacao transferencia = new Operacao(joaoConta, mariaConta, 300, TipoOperacao.TRANSFERENCIA);

        double saldoEsperadoJoao = 200;
        double saldoEsperadoMaria = 0;
        double saldoAtualJoao;
        double saldoAtualMaria;

        joao.getConta().deposito(operacaoDepositoJoao);
        joao.getConta().Transferencia(maria.getConta(), transferencia);

        saldoAtualJoao = joao.getConta().getSaldo();
        saldoAtualMaria = maria.getConta().getSaldo();

        assertEquals(saldoEsperadoJoao, saldoAtualJoao, 0.00001);
        assertEquals(saldoEsperadoMaria, saldoAtualMaria, 0.00001);
    }
}

File: src/test/kotlin/com/github/thaynarasilvapinto/simuladorbanco/ContaDonainTest.kt
package com.github.thaynarasilvapinto.simuladorbanco

import com.github.thaynarasilvapinto.simuladorbanco.domain.Cliente
import com.github.thaynarasilvapinto.simuladorbanco.domain.Conta
import com.github.thaynarasilvapinto.simuladorbanco.domain.Operacao
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.rules.ExpectedException




class ContaDonainTest {

    @get:Rule
    open var thrown = ExpectedException.none()

    lateinit var joao: Cliente
    lateinit var maria: Cliente
    lateinit var joaoConta: Conta
    lateinit var mariaConta: Conta

    lateinit var operacaoDepositoJoao: Operacao
    lateinit var operacaoSaqueJoao: Operacao
    lateinit var operacaoEfetuaTransferencia: Operacao
    lateinit var operacaorecebeTransferencia: Operacao

    @Before
    fun setUp() {
        mariaConta = Conta(saldo = 0.00)
        joaoConta = Conta(saldo = 0.00)
        this.joao = Cliente(nome = "João Pedro da Silva", cpf = "151.425.426-75", conta = joaoConta)
        this.maria = Cliente(nome = "Maria Abadia de Oliveira", cpf = "177.082.896-67", conta = mariaConta)


        this.operacaoDepositoJoao = Operacao(
                contaOrigem = joaoConta,
                contaDestino = joaoConta,
                valorOperacao = 200.00,
                tipoOperacao = Operacao.TipoOperacao.DEPOSITO)
        this.operacaoSaqueJoao = Operacao(
                contaOrigem = joaoConta,
                contaDestino = joaoConta,
                valorOperacao = 100.00,
                tipoOperacao = Operacao.TipoOperacao.SAQUE)
        this.operacaoEfetuaTransferencia = Operacao(
                contaOrigem = joaoConta,
                contaDestino = mariaConta,
                valorOperacao = 100.00,
                tipoOperacao = Operacao.TipoOperacao.TRANSFERENCIA)
        this.operacaorecebeTransferencia = Operacao(
                contaOrigem = joaoConta,
                contaDestino = mariaConta,
                valorOperacao = 100.00,
                tipoOperacao = Operacao.TipoOperacao.RECEBIMENTO_TRANSFERENCIA)
    }

    @Test
    fun `must make deposit`() {
        joao.conta.deposito(operacaoDepositoJoao)

        val saldoEsperado = 200.00
        val saldoAtual = joao.conta.saldo

        assertEquals(saldoEsperado, saldoAtual, 0.00001)
    }

    @Test
    fun `must make withdrawal`() {
        val saldoEsperado = 100.00
        val saldoAtual: Double

        joao.conta.deposito(operacaoDepositoJoao)
        joao.conta.saque(operacaoSaqueJoao)

        saldoAtual = joao.conta.saldo

        assertEquals(saldoEsperado, saldoAtual, 0.00001)
    }


    @Test
    fun `must make transference`() {
        val saldoEsperadoJoao = 100.0
        val saldoEsperadoMaria = 100.0
        val saldoAtualJoao: Double
        val saldoAtualMaria: Double
        joao.conta.saldo = 200.00
        joao.conta.efetuarTrasferencia(operacaoEfetuaTransferencia)
        maria.conta.recebimentoTransferencia(operacaorecebeTransferencia)

        assertEquals(saldoEsperadoJoao, joao.conta.saldo, 0.00001)
        assertEquals(saldoEsperadoMaria, maria.conta.saldo, 0.00001)
    }

    @Test
    fun `must not withdraw a value greater than the balance`(){
        thrown.expect(Exception::class.java)
        thrown.expectMessage("Saldo Insuficiente")
        joao.conta.saque(operacaoSaqueJoao)
    }

    @Test
    fun `must not make transference a value greater than the balance`(){
        thrown.expect(Exception::class.java)
        thrown.expectMessage("Saldo Insuficiente")
        joao.conta.efetuarTrasferencia(operacaoEfetuaTransferencia)
        maria.conta.recebimentoTransferencia(operacaorecebeTransferencia)
    }
}
